File: agent/macos/Sources/FenrirAgent/LocalControlChannel.swift
import Foundation

/// Unix-domain socket IPC server allowing a local fenrirctl CLI to query and control the daemon.
class LocalControlChannel {
    private let socketPath = "/var/run/fenrir/fenrir.sock"
    private var serverSocket: Int32 = -1
    private var handlers: [String: (([String: String]) -> [String: Any])] = [:]
    private let queue = DispatchQueue(label: "com.fenrir.ipc", attributes: .concurrent)

    func register(action: String, handler: @escaping ([String: String]) -> [String: Any]) {
        handlers[action.lowercased()] = handler
    }

    func start() {
        try? FileManager.default.createDirectory(atPath: "/var/run/fenrir", withIntermediateDirectories: true, attributes: nil)
        unlink(socketPath)

        serverSocket = socket(AF_UNIX, SOCK_STREAM, 0)
        guard serverSocket >= 0 else {
            print("[IPC] Failed to create socket")
            return
        }

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)
        withUnsafeMutablePointer(to: &addr.sun_path.0) { ptr in
            _ = socketPath.withCString { strncpy(ptr, $0, MemoryLayout.size(ofValue: addr.sun_path)) }
        }

        let bindResult = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                bind(serverSocket, $0, socklen_t(MemoryLayout<sockaddr_un>.size))
            }
        }

        guard bindResult == 0 else {
            print("[IPC] Failed to bind socket: \(String(cString: strerror(errno)))")
            return
        }

        chmod(socketPath, 0o600)
        listen(serverSocket, 5)
        print("[IPC] LocalControlChannel listening on \(socketPath)")

        queue.async { [weak self] in self?.acceptLoop() }
    }

    func stop() {
        if serverSocket >= 0 {
            close(serverSocket)
            serverSocket = -1
            unlink(socketPath)
        }
    }

    private func acceptLoop() {
        while serverSocket >= 0 {
            let clientFd = accept(serverSocket, nil, nil)
            guard clientFd >= 0 else { break }
            queue.async { [weak self] in self?.handleClient(fd: clientFd) }
        }
    }

    private func handleClient(fd: Int32) {
        defer { close(fd) }

        var buffer = [UInt8](repeating: 0, count: 4096)
        let bytesRead = read(fd, &buffer, buffer.count - 1)
        guard bytesRead > 0 else { return }

        let raw = Data(buffer.prefix(bytesRead))
        guard let cmd = try? JSONSerialization.jsonObject(with: raw) as? [String: Any],
              let action = cmd["action"] as? String else {
            send(fd: fd, response: ["success": false, "message": "invalid command"])
            return
        }

        let args = cmd["args"] as? [String: String] ?? [:]
        if let handler = handlers[action.lowercased()] {
            let result = handler(args)
            send(fd: fd, response: result)
        } else {
            send(fd: fd, response: ["success": false, "message": "unknown action: \(action)"])
        }
    }

    private func send(fd: Int32, response: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: response),
              let str = String(data: data, encoding: .utf8) else { return }
        let bytes = Array((str + "\n").utf8)
        write(fd, bytes, bytes.count)
    }
}

File: agent/linux/pkg/ipc/LocalControlChannel.go
package ipc

import (
	"encoding/json"
	"fmt"
	"log"
	"net"
	"os"
	"strings"
)

const socketPath = "/var/run/fenrir/fenrir.sock"

type Command struct {
	Action string            `json:"action"`
	Args   map[string]string `json:"args,omitempty"`
}

type CommandResponse struct {
	Success bool        `json:"success"`
	Message string      `json:"message,omitempty"`
	Data    interface{} `json:"data,omitempty"`
}

// CommandHandler is a callback that processes a received command.
type CommandHandler func(cmd Command) CommandResponse

type LocalControlChannel struct {
	listener net.Listener
	handlers map[string]CommandHandler
}

func NewLocalControlChannel() *LocalControlChannel {
	return &LocalControlChannel{
		handlers: make(map[string]CommandHandler),
	}
}

// Register adds a handler for a named action (e.g., "status", "quarantine-list").
func (l *LocalControlChannel) Register(action string, handler CommandHandler) {
	l.handlers[strings.ToLower(action)] = handler
}

func (l *LocalControlChannel) Start() error {
	if err := os.MkdirAll("/var/run/fenrir", 0700); err != nil {
		return fmt.Errorf("create socket dir: %w", err)
	}
	// Remove stale socket
	os.Remove(socketPath)

	ln, err := net.Listen("unix", socketPath)
	if err != nil {
		return fmt.Errorf("listen unix socket: %w", err)
	}
	// Only root should access the control socket
	if err := os.Chmod(socketPath, 0600); err != nil {
		log.Printf("[IPC] chmod warning: %v", err)
	}

	l.listener = ln
	log.Printf("[IPC] LocalControlChannel listening on %s", socketPath)

	go l.accept()
	return nil
}

func (l *LocalControlChannel) Stop() {
	if l.listener != nil {
		l.listener.Close()
		os.Remove(socketPath)
	}
}

func (l *LocalControlChannel) accept() {
	for {
		conn, err := l.listener.Accept()
		if err != nil {
			return // listener was closed
		}
		go l.handle(conn)
	}
}

func (l *LocalControlChannel) handle(conn net.Conn) {
	defer conn.Close()

	var cmd Command
	if err := json.NewDecoder(conn).Decode(&cmd); err != nil {
		l.reply(conn, CommandResponse{Success: false, Message: "invalid JSON"})
		return
	}

	action := strings.ToLower(cmd.Action)
	handler, ok := l.handlers[action]
	if !ok {
		l.reply(conn, CommandResponse{Success: false, Message: fmt.Sprintf("unknown action: %s", action)})
		return
	}

	resp := handler(cmd)
	l.reply(conn, resp)
}

func (l *LocalControlChannel) reply(conn net.Conn, resp CommandResponse) {
	data, _ := json.Marshal(resp)
	conn.Write(append(data, '\n'))
}
